# 演習 1-1  
PythonのリストとNumPy配列に対する**加算後のスライス処理**の実行速度を比較し、結果に基づいて考察を行え。

## 実験の目的  
PythonのリストとNumPy配列における処理速度の違いを、**処理の繰り返しによって明確に体感する**ことを目的とする。<br>
特に、処理単位を「加算処理＋スライス処理」とし、その繰り返し回数を変化させることで、処理性能の差がどのように顕在化するかを観察し、NumPyの高速性の要因を考察する。

## 実施条件
- 対象とする配列の要素数は **10万要素** とし、**整数型（int32）**とする。
- 処理単位は、以下の2操作を連続して1回とする：
  - 配列全体の各要素に `+5` を加える加算処理
  - インデックス範囲 `[1000:2000]` を取り出すスライス処理
- 上記の処理単位を、**10回・100回・1000回** の3通りで繰り返し実行する。
- 各条件について、**処理全体を5回測定し、そのうち処理時間が短い上位3回の平均**を計測結果とする。
- PythonのリストおよびNumPy配列のそれぞれについて同様の試行を行うこと。
- 以上により、**3通りの試行回数 × 2種類のデータ構造 = 計6件の測定結果**を得ること。
- 測定結果に基づき、**処理速度の違いや、試行回数の増加によって差が顕著になる様子、およびその要因について考察を行え。**

## 提出物(最後のmarkdownのセルに記述する)
- 各条件における処理時間の記録（合計6件）と比較
- 処理性能の差についての考察（原因や仕組みに対する仮説を含めること）

## 参考
練習1-3-9のコードで処理速度を計測する方法を提示している

## 補足(並べ替えについて)
Pythonのリストには並べ替えの関数が用意されている。<br>
この機能を使って、計測時間をリストに格納してからsort関数を利用することにより、処理時間の短い順に得ることができる。

In [2]:
# 並べ替えの例
a = [7,6,9,1,3,5,2,4,8]
print(a)

a.sort() # 並べ替え
print(a)

[7, 6, 9, 1, 3, 5, 2, 4, 8]
[1, 2, 3, 4, 5, 6, 7, 8, 9]


## データ準備
* 要素数は10万
* 整数（0,1,2,....,99999)
* Numpy配列は`int32`型で作成する

In [3]:
# ステップ１のコード
import numpy as np

# Pythonリストの作成
py_list = list(range(100000))

# NumPy配列の作成（int32型で明示）
np_array = np.array(py_list, dtype=np.int32)

print("データの準備が完了しました。")

データの準備が完了しました。


## 関数定義
* Pythonのリストに各要素に対して5加算し、スライスする処理を指定回繰り返す関数 `process_py_list`
* NumPy配列に各要素に対して5加算し、スラうすする処理を指定回数繰り返す関数 `process_np_array`

を次のコードセルに定義する<br>
関数内のpassを削除して、適切な処理を追加する

In [4]:
# Pythonのリストに各要素に対して5加算し、スライスする処理を指定回繰り返す関数
# 引数: pl - Pythonリスト, n - 繰り返し回数
# 戻り値: なし
def process_py_list(pl, n):
    for _ in range(n):
        for j in range(len(pl)):
            pl[j] += 5
        pl[0:n]

# NumPy配列に各要素に対して5加算し、スライスする処理を指定回繰り返す関数
# 引数: na - NumPy配列, n - 繰り返し回数
# 戻り値: なし
def process_np_array(na, n):
    for _ in range(n):
        na += 5
        na[0:n]

## 関数テスト
関数が正しく動作しているか確認する<br>
（試しに１回処理して時間を計測している）

In [5]:
import time

py_start = time.perf_counter()
process_py_list(py_list, 1) # 1回処理をする
py_time = time.perf_counter() - py_start
print(f"Pythonリスト処理時間: {py_time:.6f}秒")

np_start = time.perf_counter()
process_np_array(np_array, 1) # 1回処理をする
np_time = time.perf_counter() - np_start
print(f"NumPy配列処理時間: {np_time:.6f}秒")

Pythonリスト処理時間: 0.014595秒
NumPy配列処理時間: 0.000986秒


以降、適宜コードセルを追加し、演習を行う。<br>
複雑で大きなコードを書くのではなく、結果が得ることが目的なので、自分が扱える程度の小さなコードを書く

---

In [7]:
count = [10, 100, 1000]
py_times = []
np_times = []
for i in range(3):
    for _ in range(5):
        py_start = time.perf_counter()
        process_py_list(py_list, count[i])
        py_time = time.perf_counter() - py_start
        py_times.append(py_time)

        np_start = time.perf_counter()
        process_np_array(np_array, count[i])
        np_time = time.perf_counter() - np_start
        np_times.append(np_time)
        

    py_times.sort()
    np_times.sort()

    py_ave = np.sum(py_times[:3])/3
    np_ave = np.sum(np_times[:3])/3

    print(f"処理単位が{count[i]}回のPythonリスト処理時間: {py_time:.6f}秒")
    print(f"処理単位が{count[i]}回のNumPy配列処理時間: {np_time:.6f}秒")

処理単位が10回のPythonリスト処理時間: 0.107051秒
処理単位が10回のNumPy配列処理時間: 0.000527秒
処理単位が100回のPythonリスト処理時間: 1.075576秒
処理単位が100回のNumPy配列処理時間: 0.004463秒
処理単位が1000回のPythonリスト処理時間: 10.900782秒
処理単位が1000回のNumPy配列処理時間: 0.044650秒


----
演習のコードはここまで

# 報告　（300字以内）
計測結果を記録し、考察をしよう。NumPyは計測結果から標準のPythonより速い／速くないのか。どうしてそのような結果になるかを講義スライドやノートブックから読み取ろう
## 計測結果
- 処理単位が10回のPythonリスト処理時間: 0.107051秒
- 処理単位が10回のNumPy配列処理時間: 0.000527秒
- 処理単位が100回のPythonリスト処理時間: 1.075576秒
- 処理単位が100回のNumPy配列処理時間: 0.004463秒
- 処理単位が1000回のPythonリスト処理時間: 10.900782秒
- 処理単位が1000回のNumPy配列処理時間: 0.044650秒

## 考察
NumPyが標準のPythonリストより高速である理由は以下の通りである。  
**内部実装の違い**: NumPyはC言語で実装されており、低レベルの最適化が施されているため、計算が効率的に行われる。

>> 計算時間の大半はPythonではなくC言語によるネイティブコードで実行されるようになり大幅に高速化する。  

[参考資料](https://ja.wikipedia.org/wiki/NumPy)  
**ベクトル化演算**: NumPyはループを使用せず、配列全体に対して一括で演算を行うベクトル化を採用している。これにより、Pythonのインタプリタによるオーバーヘッドが削減される。  
**データ型の統一**: NumPy配列は固定されたデータ型（例: int32）を持つため、メモリ効率が良く、キャッシュの利用効率も向上する。一方、Pythonリストは異なる型を格納できるため、処理が遅くなる。  

以上の理由から、NumPyは大規模データの処理において特に有利である。